[gui2] Language Server binary protocol, and loading of visualization data

somebody1234 opened the pull request:
### Pull Request Description
- Depends on #7773.

- Implements binary WebSocket protocol (data protocol)
- Performs some editor initialization (the bare minimum so that visualizations work)
- Adds event handlers to receive visualization data updates

### Important Notes
None

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] ~~Unit tests have been written where possible.~~
  - [x] ~~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~~

Frizi committed:
add vue ide project files

Frizi committed:
rename gui2 folder

Frizi committed:
cleanup

Frizi committed:
cleanup

Frizi committed:
do not react to keypress when editing text

Frizi committed:
update codeowners

somebody1234 committed:
wip: (broken) begin porting visualizations to vue

somebody1234 committed:
Restore circular menu from other branch; move visualizations to separate directory

somebody1234 committed:
Use `Visualization` component in visualizations; try to figure out how to dynamically import components

somebody1234 committed:
Remove scoped styles

Vue supports `style scoped`.

somebody1234 committed:
Merge branch 'develop' into wip/sb/gui2-viz-vue-port

somebody1234 committed:
Add fullscreen button; begin fullscreen functionality

somebody1234 committed:
Fix fullscreen visualization

somebody1234 committed:
Add width/height/fullscreen props to all visualizations

somebody1234 committed:
Make visualization ref typesafe in `GraphNode`

somebody1234 committed:
Hide visualization by default

somebody1234 committed:
Add `undefined` to width/height prop

somebody1234 committed:
Transparently pass through props

somebody1234 committed:
Add `visualizationTypes` ref

somebody1234 committed:
Implement visualization selector

somebody1234 committed:
Fix caching visualizations

somebody1234 committed:
Change how visualizations are registered (tentative)

somebody1234 committed:
Fix some lints

somebody1234 committed:
Switch to new visualization format

somebody1234 committed:
Merge branch 'develop' into wip/sb/gui2-viz-vue-port

somebody1234 committed:
Visualization store

somebody1234 committed:
Add default value for image visualization

somebody1234 committed:
Move visualization names and input types to script block

somebody1234 committed:
Add support for viz style and script depenndencies

somebody1234 committed:
Remove style dependencies

somebody1234 committed:
Fix scatterplot visualization

The y-axis shows "200m"-"700m" for some reason

somebody1234 committed:
Begin porting geo map; switch to URL imports

somebody1234 committed:
Continue porting GeoMap visualization

somebody1234 committed:
Merge branch 'develop' into wip/sb/gui2-viz-vue-port

somebody1234 committed:
Add 'Error' and 'JSON' visualizations

somebody1234 committed:
remove debug code

somebody1234 committed:
Fixes for scatterplot visualization

somebody1234 committed:
Move icons to monolithic SVG; minor bug fixes

somebody1234 committed:
Dynamically compile visualizations

somebody1234 committed:
Fix CSS scoping - template scoping is still broken

somebody1234 committed:
Extract viz compiler to Web Worker; fix template scopeId not being added

somebody1234 committed:
Implement module cache

somebody1234 committed:
Minor tidying up; change one FIXME to TODO

somebody1234 committed:
Stricter typechecking

somebody1234 committed:
Continue fixing Geo Map

somebody1234 committed:
Switch from data URLs to object URLs

somebody1234 committed:
Finish porting GeoMap visualization; restore script loading for non-ESM scripts

somebody1234 committed:
Fix `z-index` - mapbox's information button has `z-index: 2`

Frizi committed:
integrate dashboard to new gui

somebody1234 committed:
Add Table visualization

somebody1234 committed:
Add space and shift+space handling for visualizations

somebody1234 committed:
Improve table visualization

somebody1234 committed:
Minor fixes to Table Visualization CSS

somebody1234 committed:
Fix minor style issues

somebody1234 committed:
Fix types in Table Visualization

somebody1234 committed:
Implement resizing

somebody1234 committed:
Add SQL visualization

somebody1234 committed:
wip: Add Heatmap Visualization

somebody1234 committed:
Add Heatmap visualization; fix points being added multiple times in Scatterplot visualization

somebody1234 committed:
Fix bugs in Heatmap visualization

somebody1234 committed:
Fix dropdown

somebody1234 committed:
Default viz width to node width

somebody1234 committed:
Properly update `nodeWidth` on resize

somebody1234 committed:
Change `nodeWidth` to `nodeSize`

somebody1234 committed:
Merge branch 'develop' into wip/sb/gui2-viz-vue-port

somebody1234 committed:
Hide circular menu by default

somebody1234 committed:
Fixes

somebody1234 committed:
Fix scatterplot brushing

somebody1234 committed:
Open circular menu on focus

somebody1234 committed:
Fix cursor when dragging

somebody1234 committed:
prettier

Frizi committed:
gateway server module code and metadata loading

Frizi committed:
use vite directly for building dashboard in gui2

Frizi committed:
fix comment

somebody1234 committed:
Add binary protocol reader and writer

somebody1234 committed:
WIP: Add bare-bones impleemntation of binary WebSocket

somebody1234 committed:
wip

somebody1234 committed:
Add more LS endpoints

somebody1234 committed:
Add viz update handler

somebody1234 committed:
Merge branch 'wip/sb/gui2-viz-vue-port' into wip/sb/connect-viz

somebody1234 committed:
fix bug

somebody1234 committed:
Try to get data server working

somebody1234 committed:
prettier

somebody1234 committed:
Continue trying to get visualizations to work

somebody1234 committed:
Switch from `Emitter` to `ObservableV2`

somebody1234 committed:
wip: More execution context initialization

somebody1234 committed:
Partially address review

somebody1234 committed:
Use `ToggleIcon` in `CircularMenu`

Spent way too long trying to get nested `v-model` to work nicely

somebody1234 committed:
Move `VisualizationContainer` back to `src/`

somebody1234 committed:
Supply visualization config by `provide` instead of props

somebody1234 committed:
Firefox scrollbar styling

somebody1234 committed:
Initial implementation of histogram visualization; reorder builtin visualizations

somebody1234 committed:
Fix histogram visualization

somebody1234 committed:
Switch from `.enter`/`.exit` to `.join`

somebody1234 committed:
Fix autosizing in Table visualization

somebody1234 committed:
wip: Histogram fixes

somebody1234 committed:
Fix scatterplot axis labels

somebody1234 committed:
Remove null checks for D3

somebody1234 committed:
Begin implementing D3 feedback

somebody1234 committed:
Simplify `useEvent` and use it

somebody1234 committed:
Continue implementing D3 feedback

somebody1234 committed:
Continue implementing D3 feedback

somebody1234 committed:
fixes

somebody1234 committed:
more fixes

somebody1234 committed:
Replace `d3Types` with `import type {}`

somebody1234 committed:
Fix zooming not working in Histogram

somebody1234 committed:
wip: Fix scatterplot

somebody1234 committed:
Fix scatterplot

somebody1234 committed:
Refactor Heatmap visualization

somebody1234 committed:
Merge branch 'wip/sb/gui2-viz-vue-port' into wip/sb/connect-viz

somebody1234 committed:
Fix binary protocol; attempt to get visualizations working

somebody1234 committed:
Add `text/` endpoints to LS API wrapper

somebody1234 committed:
Add `vcs/` endpoints to LS API wrapper

somebody1234 committed:
Make websocket calls match old GUI

somebody1234 committed:
prettier

somebody1234 committed:
Add Language Server error codes

somebody1234 committed:
Fix casing of enum members

somebody1234 committed:
Finally fix visualization data retrieval

somebody1234 committed:
Address review

somebody1234 committed:
Add docs for visualization compiler

somebody1234 committed:
Integrate icons for builtin viz into `icons.svg`

somebody1234 committed:
Merge branch 'develop' into wip/sb/gui2-viz-vue-port

somebody1234 committed:
Merge branch 'wip/sb/gui2-viz-vue-port' into wip/sb/connect-viz

somebody1234 committed:
Fix type errors

somebody1234 committed:
Fix bugs

somebody1234 committed:
prettier

somebody1234 committed:
Add license for websocket module

somebody1234 committed:
Remove spurious changes

somebody1234 committed:
prettier

somebody1234 committed:
Merge branch 'wip/sb/gui2-viz-vue-port' into wip/sb/connect-viz

somebody1234 commented:
@Frizi I *think* this is more or less ready for review too.

somebody1234 committed:
Remove `expressionUpdates` listener from gui

somebody1234 committed:
Attempt to fix ci

somebody1234 committed:
Merge branch 'develop' into wip/sb/connect-viz

somebody1234 commented:
:sob: the viz shows below other nodes

somebody1234 committed:
Fixes

Frizi commented on app/gui2/src/components/GraphNode.vue:
```diff
@@ -262,18 +297,55 @@ function handleClick(e: PointerEvent) {
 
-const visualizationType = ref('Scatterplot')
-const visualization = shallowRef<Visualization>()
+watch(isVisualizationVisible, async (visible) => {
+  if (!visible) {
+    return
+  }
+  // TODO: should the old vis be detached?
+  await props.languageServer.attachVisualization(
+    visualizationId.value,
+    props.node.rootSpan.id,
+    visualizationConfiguration.value ?? {
+      executionContextId: props.executionContextId,
+      ...DEFAULT_VISUALIZATION_CONFIGURATION,
+    },
+  )
+})
```
Detaching is important. Otherwise we will be constantly adding work on the language server side. You can us e`onCleanup` of `watchEffect` here to make it a bit simpler. Just make sure to call it before first `await`.

Frizi replied:
Detaching is important. Otherwise we will be constantly adding work on the language server side. You could use `onCleanup` of `watchEffect` here to make it a bit simpler. Just make sure to call it before first `await`.

Though it probably should be mostly handled elsewhere anyway (see next comments).

Frizi commented on app/gui2/shared/websocket.ts:
Why is this file necessary? Can't we use `lib0/websocket` directly instead?

somebody1234 replied:
it doesn't support sending binary messages, and uses a custom JSON-based ping/pong (which the server must support, and i believe our servers don't support it)

(see https://github.com/dmonad/lib0/issues/73)

Frizi commented on app/gui2/src/components/GraphEditor.vue:
```diff
@@ -106,6 +137,10 @@ function moveNode(id: ExprId, delta: Vec2) {
         v-for="[id, node] in graphStore.nodes"
         :key="id"
         :node="node"
+        :main-module="mainModule"
+        :execution-context-id="executionContextId"
+        :language-server="projectStore.lsRpcConnection"
+        :data-server="projectStore.dataConnection"
```
Inconsistent prop casing. Also I'm not a fan of passing `dataServer` and related things to the nodes here. The low level part of managing the execution context and visualization messages should be moved out of the components.

somebody1234 replied:
i was assuming `v-bind` props were supposed to stay as `kebab-case` - can be changed 

re: the low level part of managing the execution context and visualization messages - yeah, this will probably be addressed in adam's PR, i'll probably wait for that to land before merging that into this PR.

i'll see what i can do before that gets merged in though

Frizi commented on app/gui2/src/components/GraphNode.vue:
```diff
@@ -262,18 +297,55 @@ function handleClick(e: PointerEvent) {
-  visualizationStore.sampleData(visualizationType.value),
-)
-const visualizationData = ref<{}>({})
+watch(visualizationConfiguration, async (config) => {
+  if (config != null && isVisualizationVisible.value) {
+    await props.languageServer.modifyVisualization(visualizationId.value, config)
+  }
+})
+
+function onVisualizationUpdate(vizUpdate: VisualizationUpdate) {
+  const json = vizUpdate.dataString()
+  const newData = json != null ? JSON.parse(json) : undefined
+  visualizationData.value = newData
+}
+
+onMounted(() => {
+  props.dataServer.on(
+    `${OutboundPayload.VISUALIZATION_UPDATE}:${visualizationId.value}`,
+    onVisualizationUpdate,
+  )
+})
+
+onUnmounted(() => {
+  props.dataServer.off(
+    `${OutboundPayload.VISUALIZATION_UPDATE}:${visualizationId.value}`,
+    onVisualizationUpdate,
+  )
+})
```
As before, I don't like that this low level communication is done directly in the vue component's code. I think here we should at most have something like:
```ts
const visualizationData = useVisualizationData(visualizationConfigRef)
```

Frizi commented on app/gui2/shared/binaryProtocol.ts:
```diff
@@ -0,0 +1,2217 @@
+// automatically generated by the FlatBuffers compiler, and modified
+// Helper classes copied from https://github.com/google/flatbuffers,
+// which is licensed under the Apache License, Version 2.0:
```
Could you describe a bit more how it was generated?
- the tool and version used (i assume flatc, but not sure which one)
- what kind modifications were made

Frizi commented on app/ide-desktop/lib/dashboard/esbuild-config.ts:
```diff
@@ -107,7 +107,9 @@ export function bundlerOptions(args: Arguments) {
-            esbuildPluginInlineImage({ extensions: ['svg'] }),
+            esbuildPluginInlineImage({
+                extensions: ['svg'],
+            }) as esbuild.Plugin,
```
That change seems unnecessary. If you have type errors, double-check if you didn't install two different esbuild versions.

Frizi commented on app/ide-desktop/lib/dashboard/src/authentication/src/error.ts:
```diff
@@ -54,7 +54,9 @@ export function getMessageOrToString<T>(error: MustNotBeKnown<T>) {
 export function render(f: (message: string) => string): toastify.UpdateOptions {
-    return { render: ({ data }) => f(getMessageOrToString(data)) }
+    return {
+        render: ({ data }: toastify.ToastContentProps<unknown>) => f(getMessageOrToString(data)),
```
Again, types should match just fine here on their own. Make sure you have the dependencies installed correctly.

![image](https://github.com/enso-org/enso/assets/919491/61126d63-2400-4f61-8f3c-c3f38c22afdc)
![image](https://github.com/enso-org/enso/assets/919491/01849b86-b188-41c0-b87c-b27d6772a952)

Frizi commented on package-lock.json:
Those lockfile changes seems very excessive. The only thing that should change is `lib0` version. Try reverting it and doing `npm install` again.

somebody1234 committed:
Address minor review issues

somebody1234 committed:
Refactor visualization updates into project store

somebody1234 committed:
Merge branch 'develop' into wip/sb/connect-viz

somebody1234 committed:
Fix bugs?

somebody1234 commented:
ok so the viz should be working (inconsistently as usual though)

... really not sure about how the execution context is currently set up. it might be nice to refactor the execution context setup in another PR?

somebody1234 committed:
Refactor execution context into project store

somebody1234 committed:
Revert changes to ide-desktop

somebody1234 committed:
Merge branch 'develop' into wip/sb/connect-viz

somebody1234 committed:
Fixes

somebody1234 committed:
Fix?

somebody1234 committed:
Fix visualization selector not closing

somebody1234 committed:
Merge branch 'develop' into wip/sb/connect-viz

farmaazon committed:
Merge remote-tracking branch 'origin/develop' into wip/sb/connect-viz

somebody1234 committed:
Fix changing visualization type

somebody1234 commented:
note: this is still *not tested* with a new build of PM, so it may be broken.

Frizi committed:
Merge branch 'develop' into wip/sb/connect-viz

Frizi committed:
More robust execution context.
- Added stateful async queue for managing visualization registration.
- Added error handling and exponential backoff to many language server requests.

Frizi committed:
visualizations in node metadata

farmaazon commented on app/gui2/shared/languageServerTypes.ts:
```diff
@@ -265,7 +265,7 @@ export interface VisualizationConfiguration {
   expression: string | MethodPointer
   /** A list of arguments to pass to the visualization expression. */
-  positionalArgumentsExpressions?: string[]
+  positionalArgumentsExpressions: string[]
```
It _is_ defined as an optional field in the document, and there is a difference between an empty array and a missing field.

Frizi replied:
I don't see any explanation of a difference between missing value and an empty array in the protocol doc. Do you know what it is?

Frizi committed:
fix node-height

somebody1234 commented on app/gui2/src/components/GraphEditor.vue:
```diff
@@ -1,18 +1,4 @@
-
 import { nodeBindings } from '@/bindings/nodeSelection'
 import CodeEditor from '@/components/CodeEditor.vue'
 import ComponentBrowser from '@/components/ComponentBrowser.vue'
```
@Frizi just want to confirm that this reordering of imports is intentional?

Frizi replied:
Yes. I found that using separate import lists keeps the IDE auto-format breaking. The only working configuration is if all imports are in the first `script` block in a file. To reduce friction, I think we should simply accept that as a rule.

somebody1234 commented on app/gui2/src/components/visualizations/LoadingVisualization.vue:
```diff
@@ -0,0 +1,43 @@
+<script lang="ts">
+export const name = 'JSON'
```
incorrect name

somebody1234 commented on app/gui2/src/components/visualizations/SQLVisualization.vue:
```diff
@@ -4,7 +4,7 @@ export const inputType = 'Standard.Database.Data.Table.Table | Standard.Database
   'Standard.Visualization.SQL.Visualization',
   'prepare_visualization',
-]
+] as const
```
are these `as const`s really needed?

Frizi replied:
It makes the type true to what's specified in `VisualizationModule`. I attempted to remove the `as any` from `builtinVisualizationImports` list, and I sort of succeeded with it, at least in IDE. Unfortunately it still breaks during external typecheck, so I've kept the any casts for now. But at least I've made some progress towards that.

somebody1234 commented on app/gui2/src/components/GraphNode.vue:
```diff
@@ -294,67 +298,62 @@ function updatePreprocessor(
+
+function setVisualizationIdentifier(id: VisualizationIdentifier) {
+  emit('setVisualizationId', id)
 }
```
idk whether a wrapper function that only contains a single function call and nothing else, is all that useful

somebody1234 commented on app/gui2/src/components/VisualizationContainer.vue:
```diff
@@ -127,8 +127,9 @@ const resizeBottomRight = usePointer((pos, _, type) => {
+              :value="config.currentType"
               @hide="isSelectorVisible = false"
-              @update:type="
+              @update:value="
```
`modelValue` + `update:modelValue` may be more conventional than `value` + `update:value`

Frizi committed:
fix loading visualization name

somebody1234 commented on app/gui2/src/util/shortcuts.ts:
```diff
@@ -317,13 +322,6 @@ function decomposeKeybindString(string: string): ModifierStringDecomposition {
   const decomposed = decomposeKeybindString(string)
-  const normalized =
-    decomposed.modifiers.length === 0
-      ? decomposed.key
-      : `${decomposed.modifiers.join('+')}+${decomposed.key}`
-  if (normalized !== string) {
-    console.warn(`Modifier string '${string}' should be '${normalized}'`)
-  }
   if (isPointer(decomposed.key)) {
```
?!

somebody1234 replied:
well, i think it's not *vital* but i think it'd be nice to sort of have builtin linting - it's true that this isn't useful for builtin shortcuts, but i think it'd be nice for user-defined shortcuts. that said, `console.warn` isn't very visible, but i don't think we have a well-defined place to put status notifications right now

Frizi replied:
This whole lint was incorrect once you used any overrides in `normalizedKeyboardSegmentLookup`. E.g. using `OsDelete` or `Space` (which I added, since it was impossible to use `' '` due to trim in decomposition logic). After all, `decomposed.key` is simply not necessarily what you actually want to write in your binding.

We could add normalization and a lint again, but it would require a bit more changes to the whole decomposition step.

somebody1234 replied:
ah yeah, good point. i think this was written before the overrides and i kinda just forgot that that would invalidate this

somebody1234 commented on app/gui2/ydoc-server/edits.ts:
```diff
@@ -7,7 +7,7 @@ import diff from 'fast-diff'
 import * as Y from 'yjs'
 import { TextEdit } from '../shared/languageServerTypes'
-import { ModuleDoc, NodeMetadata, decodeRange } from '../shared/yjsModel'
+import { ModuleDoc, NodeMetadata, VisualizationMetadata, decodeRange } from '../shared/yjsModel'
```
should `VisualizationMetadata` be a `type` import?

somebody1234 commented on app/gui2/ydoc-server/languageServerSession.ts:
```diff
@@ -362,14 +367,16 @@ class ModulePersistence extends ObservableV2<{ removed: () => void }> {
 
     this.changeState(LsSyncState.WritingFile)
+
+    const execute = true // contentDelta != null || idMapKeys != null
```
commented code

farmaazon commented on app/gui2/src/stores/project.ts:
```diff
@@ -103,31 +126,192 @@ async function initializeDataConnection(clientId: Uuid, url: string) {
+      thisArgumentExpression,
+      positionalArgumentsExpressions: positionalArgumentsExpressions ?? [],
+    })
+  }
```
I think this should not be part of public API. You always want to have just one "call" frame on the bottom of the stack - so it should be done in `create` method.

farmaazon commented on app/gui2/src/util/net.ts:
```diff
@@ -0,0 +1,106 @@
+    ) {
+      return result
+    }
+    await wait(delay)
```
Shan't we print an error log here?

Frizi replied:
This is a low-level utility. The adequate logs are already printed by its users whenever it makes sense.

farmaazon commented on app/gui2/src/util/net.ts:
```diff
@@ -0,0 +1,106 @@
+ * takes the current state and produces a promise to the transformed state. Each task waits for the
+ * previous task to finish before starting.
+ */
+export class AsyncQueue<State> {
+  lastTask: Promise<State>
+  taskRunning = false
+  queuedTasks: QueueTask<State>[] = []
+
+  constructor(initTask: Promise<State>) {
+    this.lastTask = initTask
+  }
+
+  private run() {
+    if (this.taskRunning) return
+    const task = this.queuedTasks.shift()
+    if (task == null) return
+    this.taskRunning = true
+    this.lastTask = this.lastTask
+      .then((state) => task(state))
+      .finally(() => {
+        this.taskRunning = false
+        this.run()
+      })
+  }
+
+  pushTask(f: QueueTask<State>) {
+    this.queuedTasks.push(f)
+    this.run()
+  }
+
+  clear() {
+    this.queuedTasks.length = 0
+  }
+}
```
I'd be nice to see tests here.

Even for exponential backoff, here however I don't know how to handle waits.

Frizi replied:
Waits should actually be pretty easily testable. It's all a `setTimeout` under the hood, and those are already [mocked by vitest](https://vitest.dev/guide/mocking.html#timers). I'll add some tests then.

farmaazon reviewed (approved):
I''d like to see some tests of

Frizi committed:
add net util tests

Frizi committed:
remove `pushCall` from execution context public API.

Frizi committed:
fix gui2 build, nicer code editor styling
